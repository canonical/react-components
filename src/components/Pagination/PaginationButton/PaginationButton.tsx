import classNames from "classnames";
import React from "react";
import type { MouseEventHandler } from "react";

export enum Label {
  Next = "Next page",
  Previous = "Previous page",
}

export type PaginationDirection = "forward" | "back";
export type Props = {
  /**
   * The direction of the pagination.
   */
  direction: PaginationDirection;
  /**
   * Whether the pagination button should be disabled.
   */
  disabled?: boolean;
  /**
   * Function to handle clicking the pagination button.
   */
  onClick: MouseEventHandler<HTMLButtonElement>;
  /**
   * Whether to show the label for button.
   */
  showLabel?: boolean;
  /**
   * Custom label for button.
   */
  label?: string;
};

const PaginationButton = ({
  direction,
  onClick,
  disabled = false,
  showLabel,
  label,
}: Props): JSX.Element => {
  const buttonLabel =
    label || (direction === "back" ? Label.Previous : Label.Next);
  return (
    <li className="p-pagination__item">
      <button
        className={classNames({
          "p-pagination__link--previous": direction === "back",
          "p-pagination__link--next": direction === "forward",
        })}
        disabled={disabled}
        onClick={onClick}
        type="button"
      >
        {direction === "forward" && showLabel && <span>{buttonLabel}</span>}
        <i className="p-icon--chevron-down">{buttonLabel}</i>
        {direction === "back" && showLabel && <span>{buttonLabel}</span>}
      </button>
    </li>
  );
};

export default PaginationButton;
