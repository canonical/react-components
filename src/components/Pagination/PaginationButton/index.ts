export { default } from "./PaginationButton";
export type { Props as PaginationButtonProps } from "./PaginationButton";
