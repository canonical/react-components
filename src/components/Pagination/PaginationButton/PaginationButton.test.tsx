import React from "react";
import { render, screen } from "@testing-library/react";

import PaginationButton from "./PaginationButton";

describe("PaginationButton", () => {
  it("should render correctly and match snapshot", () => {
    render(<PaginationButton direction="back" onClick={() => {}} />);
    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("should contain default label after the back icon", () => {
    render(<PaginationButton direction="back" onClick={() => {}} showLabel />);
    const customLabel = screen
      .getByRole("button", { name: /Previous page/ })
      // eslint-disable-next-line testing-library/no-node-access
      .querySelector("span");
    // eslint-disable-next-line testing-library/no-node-access
    const icon = document.querySelector(".p-icon--chevron-down");
    expect(customLabel).toHaveTextContent("Previous page");
    expect(
      customLabel &&
        icon &&
        customLabel.compareDocumentPosition(icon) &
          Node.DOCUMENT_POSITION_PRECEDING
    ).toBeTruthy();
  });

  it("should contain custom label before the forward icon", () => {
    render(
      <PaginationButton
        direction="forward"
        onClick={() => {}}
        showLabel
        label="Custom label"
      />
    );
    const customLabel = screen
      .getByRole("button", { name: /Custom label/ })
      // eslint-disable-next-line testing-library/no-node-access
      .querySelector("span");
    // eslint-disable-next-line testing-library/no-node-access
    const icon = document.querySelector(".p-icon--chevron-down");
    expect(customLabel).toHaveTextContent("Custom label");
    expect(
      customLabel &&
        icon &&
        customLabel.compareDocumentPosition(icon) &
          Node.DOCUMENT_POSITION_FOLLOWING
    ).toBeTruthy();
  });
});
