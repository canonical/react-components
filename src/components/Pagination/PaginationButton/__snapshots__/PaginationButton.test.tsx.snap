// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationButton should render correctly and match snapshot 1`] = `
<button
  class="p-pagination__link--previous"
  type="button"
>
  <i
    class="p-icon--chevron-down"
  >
    Previous page
  </i>
</button>
`;
